### PR TITLE
Added support for Moldova

### PR DESCRIPTION
**Pull request by vladimirnani**
Adds few moldavian specific fields. 

- Moldavian company types
- Moldavian IDNO (Company registration number)
- Moldavian License Plate

Closes: #269 

**All Changes**
- [x] Add an entry to the docs/changelog.rst describing the change.
- [x] Add an entry for your name in the docs/authors.rst file if it's not
      already there.
- [x] Adjust your imports to a standard form by running this command:
      `isort --recursive --line-width 120 localflavor tests`

**New Fields Only**
- [x] Prefix the country code to all fields.
- [x] Field names should be easily understood by developers from the target
      localflavor country. This means that English translations are usually
      not the best name unless it's for something standard like postal code,
      tax / VAT ID etc.
- [x] Prefer '<country code>PostalCodeField' for postal codes as it's
      international English; ZipCode is a term specific to the United
      States postal system.
- [x] Add meaningful tests. 100% test coverage is not required but all
      validation edge cases should be covered.
- [x] Add `.. versionadded:: <next-version>` comment markers to new
      localflavors.
- [x] Add documentation for all fields.

### CODE REVIEW

File: localflavor/md/forms.py [hunk 1]
```diff
@@ -0,0 +1,58 @@
+# -*- coding: utf-8 -*-
+from __future__ import unicode_literals
+
+from django import forms
+
+from localflavor.md.choices import COMPANY_TYPES_CHOICES
+from localflavor.md.validators import MDIDNOFieldValidator, MDLicensePlateValidator
```
Comment (benkonrath) [inline localflavor/md/forms.py]: Just to be consistent with other flavors, you should use relative imports for code in this flavor (e.g. `from .choices import REGION_CHOICES`). You should change this in the other files too.

File: localflavor/md/forms.py [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+
+class MDRegionSelect(forms.Select):
+    """
+    A Select widget that uses a list of Moldavian regions as its choices.
+
+    .. versionadded:: 1.6
+    """
+
+    def __init__(self, attrs=None):
+        from .choices import REGION_CHOICES
```
Comment (benkonrath) [inline localflavor/md/forms.py]: This doesn't need to be a local import.

File: localflavor/md/validators.py [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+    """
+
+    error_message = _('Enter a valid IDNO number.')
+
+    def __init__(self):
+        super(MDIDNOFieldValidator, self).__init__(regex='^\d{13}$',
+                                                   message=self.error_message)
+
+    def __call__(self, value):
+        super(MDIDNOFieldValidator, self).__call__(value)
```
Comment (benkonrath) [inline localflavor/md/validators.py]: This doesn't do anything so it should be removed.

File: localflavor/md/validators.py [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+    Validation for Moldavian IDNO.
+
+    .. versionadded:: 1.6
+    """
+
+    error_message = _('Enter a valid IDNO number.')
+
+    def __init__(self):
+        super(MDIDNOFieldValidator, self).__init__(regex='^\d{13}$',
+                                                   message=self.error_message)
```
Comment (benkonrath) [inline localflavor/md/validators.py]: It's better to just override the `regex` and `message` variables from the `RegexValidator` instead of setting them manually in `__init__`.

I read a translation of this site that says an IDNO is an alpha numeric value. Is the regex correct (only digits)?
https://servicii.gov.md/ServiceDetails.aspx?id=5b54755b-174c-4576-a8a4-5df336eca90d

File: localflavor/md/validators.py [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+
+    def __call__(self, value):
+        super(MDIDNOFieldValidator, self).__call__(value)
+
+
+# https://en.wikipedia.org/wiki/Vehicle_registration_plates_of_Moldova
```
Comment (benkonrath) [inline localflavor/md/validators.py]: It would be useful to have this link in the doc string.

File: localflavor/md/validators.py [hunk 6]
```diff
@@ -0,0 +1,107 @@
...
+    Validation for Moldavian License Plates.
+
+    .. versionadded:: 1.6
+    """
+
+    error_message = _('Enter a valid license plate.')
+
+    def __init__(self):
+        super(MDLicensePlateValidator, self).__init__(regex='^\d{9}$',
+                                                      message=self.error_message)
```
Comment (benkonrath) [inline localflavor/md/validators.py]: Same as above; override `regex` and `message` in `RegexValidator`.

File: localflavor/md/validators.py [hunk 7]
```diff
@@ -0,0 +1,107 @@
...
+            self._is_diplomatic_format(value) or
+            self._is_police_format(value) or
+            self._is_foreign_format(value) or
+            self._is_state_security_format(value) or
+            self._is_gov_format(value)
+        )
+
+    @staticmethod
+    def _is_old_format(value):
+        from localflavor.md.choices import REGION_CHOICES
```
Comment (benkonrath) [inline localflavor/md/validators.py]: These should be imported at the top (e.g. `from .choices import REGION_CHOICES`). The same applies to the other local imports.

**Comment by benkonrath** [comment]
@vladimirnani I'm pushing this forward to a 2.x release because I'm going to release 1.6 soon. You'll have to update the `versionaddded` strings when you get back to this. Thanks.

**Comment by benkonrath** [comment]
@vladimirnani I'm hoping to release 2.0 shortly (see #322). If you have time to make the requested changes we can try to get this into the 2.0 release.

**Comment by benkonrath** [comment]
Let's target the 2.1 release for this so that the strings have a chance to be translated correctly.

File: localflavor/md/validators.py (hunk 4 above)
Comment (vladimirnani) [reply localflavor/md/validators.py]: For sure IDNO is only numeric, its a company id, not sure about IDNP (personal).
In our case its only companies ids

**Comment by vladimirnani** [comment]
Thanks for the feedback. 
Fixed code comments, included translation of company types in russian. Having troubles finding correct translation for company types in english. Also for example SRL stands for '**S**ocietatea cu **R**aspundere **L**imitata' so im not sure if there is sense to have english message ids instead.

File: docs/changelog.rst [hunk 8]
```diff
@@ -110,6 +110,9 @@ New flavors:
 - Added local flavor for Cuba
   (`gh-292 <https://github.com/django/django-localflavor/pull/292>`_).
 
+- Added local flavor for Moldova
+  (`gh-309 <https://github.com/django/django-localflavor/pull/309>`_)
+
```
Comment (benkonrath) [inline docs/changelog.rst]: This entry should be in the 2.1 changelog section.

File: localflavor/md/forms.py [hunk 9]
```diff
@@ -0,0 +1,57 @@
...
+
+from .choices import COMPANY_TYPES_CHOICES, REGION_CHOICES_2002_2015
+from .validators import MDIDNOFieldValidator, MDLicensePlateValidator
+
+
+class MDIDNOField(forms.CharField):
+    """
+    A form field for the Moldavian company identification number (IDNO).
+
+    .. versionadded:: 1.6
```
Comment (benkonrath) [inline localflavor/md/forms.py]: 1.6 -> 2.1

Comment (benkonrath) [reply localflavor/md/forms.py]: Same applies to all of the 1.6 references.

File: localflavor/md/validators.py (hunk 4 above)
Comment (benkonrath) [reply localflavor/md/validators.py]: Ok, the numberic regex is fine. I trust you know how it's used.

I still think you should override the `regex` and `message` variables from the `RegexValidator` instead of setting them manually in `__init__`. For example:

```
class MDIDNOFieldValidator(RegexValidator):
     regex = '^\d{13}$'
     message = _('Enter a valid IDNO number.')
```

File: localflavor/md/validators.py (hunk 3 above)
Comment (benkonrath) [reply localflavor/md/validators.py]: This still needs to be removed.

File: localflavor/md/validators.py [hunk 10]
```diff
@@ -0,0 +1,111 @@
...
+
+    .. versionadded:: 1.6
+    """
+
+    error_message = _('Enter a valid license plate.')
+    # regex = '^\d{9}$'
+    # message = error_message
+
+    def __init__(self):
+        super(MDLicensePlateValidator, self).__init__(regex='^\d{9}$',
```
Comment (benkonrath) [inline localflavor/md/validators.py]: Same comment as above: you should override `regex` and `message` instead of setting them in `__init__`. Or was that not working?

The commented code should be removed as well.

**Comment by vladimirnani** [comment]
- [ ] Adapted versions and change log
- [ ] Removed localization
- [ ] Field initializers instead of ctor

File: localflavor/md/validators.py [hunk 11]
```diff
@@ -0,0 +1,98 @@
...
+    message = error_message
+
+    def __call__(self, value):
+        value = value.upper()
+        if not self._is_valid(value):
+            raise ValidationError(self.error_message)
+
+    def _is_valid(self, value):
+        return (
+            self._is_old_format(value) or
```
Comment (claudep) [inline localflavor/md/validators.py]: Small possible improvement with `any()`.

**Comment by claudep** [comment]
Great achievement, congrats!

File: localflavor/md/validators.py (hunk 11 above)
Comment (benkonrath) [reply localflavor/md/validators.py]: Good suggestion (sorry for jumping the gun).

@vladimirnani  Is it possible for you to make a small PR to update this?

Comment (vladimirnani) [reply localflavor/md/validators.py]: Sure, thanks for feadback @claudep

Comment (vladimirnani) [reply localflavor/md/validators.py]: @claudep take a look please, #344

**Comment by benkonrath** [comment]
@vladimirnani I just wanted to say thanks for putting in the extra effort to get this PR done in time for the next release. If I remember correctly, the idea for this localflavor came up during the Django Under The Hood sprints in 2016. I'm super excited to finally have this in localflavor. A big thanks! :+1:

**Comment by vladimirnani** [comment]
It took too long, but thanks for the patience and help!

**Comment by benkonrath** [comment]
It took the time it needed to take. :-)
